### ISSUE TITLE
Low Scores Compared to nltk Implementation

### ISSUE THREAD

**Issue by mertozlutiras-hyrd** [facebookresearch/faiss#2492]
# Summary

I'm trying to cluster 55K text embeddings (each has a dimension of 768) into 1K+ clusters. 

These clusters should be optimized based on cosine similarity of embedding vectors.

I'm comparing my models' performance based on silhouette scores (https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html) and sum of squred errors.

When I used nltk implementation of KMeans clustering (https://www.nltk.org/_modules/nltk/cluster/kmeans.html), training takes 8+ hours. And I get silhouette scores around 0.48, which is satisfactory.

Since training for 8+ hours for every experiment is not feasible for me, I'd been looking for alternatives. I used faiss' KMeans implementation, which works wonders in terms of training time (5mins). However, silhouette scores I get are far from acceptable. I get scores around 0.08.

I realized faiss uses Euclidan distance instead of cosine similarity. In a github discussion a user reported if normalized embeddings are used, training both with Euclidean distance and cosine similarity should produce same results. I've also used normalization and retested faiss scores. However, results are still the same.

Do you have any suggestion on how to produce same results with faiss as nltk?

Installed from: anaconda

Running on:
- [ ] CPU
- [ X ] GPU

Interface: 
- [ ] C++
- [ X ] Python

# Reproduction instructions

Here:
```

  class FaissKMeans:
  
      def __init__(self, n_clusters=8, n_init=10, max_iter=300):
          self.n_clusters = n_clusters
          self.n_init = n_init
          self.max_iter = max_iter
          self.kmeans = None
          self.cluster_centers_ = None
          self.inertia_ = None
          self.silhouette = None
      def fit(self, data_dict):
          X = np.array(data_dict['embeddings']).astype('float32')
          faiss.normalize_L2(X)
          self.kmeans = faiss.Kmeans(d=X.shape[1],
                                     k=self.n_clusters,
                                     niter=self.max_iter,
                                     nredo=self.n_init,
                                     gpu = True,
                                     verbose = True)
                                     #index = quantizer)
          self.kmeans.train(X)
          self.cluster_centers_ = self.kmeans.centroids
          self.inertia_ = self.kmeans.obj[-1]
  
      def predict(self, data_dict):
          X = np.array(data_dict['embeddings']).astype('float32')
  
          assigned_clusters = [float(element) for element in self.kmeans.index.search(X, 1)[1]]
          self.silhouette = silhouette_score(X, assigned_clusters)
          return assigned_clusters
  
  
  ## TESTING FOR DIFFERENT PARAMETERS
  
  list_n_clusters = [(i+1)*500 for i in range(10)]
  
  results = []
  
  for n_clusters in list_n_clusters:
    kmeans_faiss = FaissKMeans(n_clusters = n_clusters, n_init=5)
    kmeans_faiss.fit(data_dict)
    predictions = kmeans_faiss.predict(data_dict)
  
    inertia = kmeans_faiss.inertia_
    silhouette = kmeans_faiss.silhouette
  
    results.append({'n_clusters': n_clusters,
                    'scores': {
                    'inertia': inertia,
                    'silhouette': silhouette
                    }
                    })

```

**Comment by mdouze** [comment]
That's interesting, there does not seem to be anything special with the NLTK implementation of k-means. It is obviously very slow because everything is implemented in pure python and without vectorization, but the algorithm is standard. There is no normalization of the centroids. 

I would use a smaller scale set, take the NLTK implementation and progressively replace them with Faiss functions. In fact the code: 
```
              for vector in vectors:
                    index = self.classify_vectorspace(vector)
``` 
is what makes it slow, it can be replaced with a single`faiss.knn` call, see https://github.com/facebookresearch/faiss/wiki/Brute-force-search-without-an-index

**Comment by mertozlutiras-hyrd** [comment]
> That's interesting, there does not seem to be anything special with the NLTK implementation of k-means. It is obviously very slow because everything is implemented in pure python and without vectorization, but the algorithm is standard. There is no normalization of the centroids.
> 
> I would use a smaller scale set, take the NLTK implementation and progressively replace them with Faiss functions. In fact the code:
> 
> ```
>               for vector in vectors:
>                     index = self.classify_vectorspace(vector)
> ```
> 
> is what makes it slow, it can be replaced with a single`faiss.knn` call, see https://github.com/facebookresearch/faiss/wiki/Brute-force-search-without-an-index

I think the issue is with the distance metric. In nltk implementation, I can directly choose the distance metric to be optimized as cosine similarity with parameter 'distance=nltk.cluster.util.cosine_distance'. However in FAISS, I think I couldn't achieve to change the distance metric to be optimized to cosine distance. I think it's still optimizing for euclidean distance.

I am not sure if I completely understood the FAISS package. I am not sure how to use the 'Index' objects. I directly used faiss.KMeans object.

I've seen this question many times appearing in the past issues, but couldn't find an elaborate description. Could you help with how to optimize the KMeans algorithm for cosine distance instead of Euclidean?

**Comment by mertozlutiras-hyrd** [comment]
> That's interesting, there does not seem to be anything special with the NLTK implementation of k-means. It is obviously very slow because everything is implemented in pure python and without vectorization, but the algorithm is standard. There is no normalization of the centroids.
> 
> I would use a smaller scale set, take the NLTK implementation and progressively replace them with Faiss functions. In fact the code:
> 
> ```
>               for vector in vectors:
>                     index = self.classify_vectorspace(vector)
> ```
> 
> is what makes it slow, it can be replaced with a single`faiss.knn` call, see https://github.com/facebookresearch/faiss/wiki/Brute-force-search-without-an-index

Can you share a documentation of the python interface? As far as I see, faiss.knn doesn't return the centroids (or cluster index), but it returns the nearest neighbor.  I need to cluster them and return a centroid index.

**Comment by mertozlutiras-hyrd** [comment]
After so much effort to make it work, I still couldn't achieve to do so.

I tried everything mentioned in the closed issues (normalization, spherical, higher max_points_per_centroid, using IndexIVFFlat with quantizer..)

I've seen this 'cosine similarity' issue coming up many times in the past and still no elaborate answer/help was provided.

Due to lack of proper python documentation, and not enough help received I gave up on faiss and I'll stick with slow nltk implementation.

@mdouze
